src/hivpy/circumcision.py

```diff
@@ -0,0 +1,182 @@
+           | ((self.year_interv + 5 <= self.date.year)
+              & (self.circ_inc_rate_year_i == 4)):
+            self.test_link_circ_prob = 0
+
+        # only begin VMMC after a specific year
+        elif self.mc_int < self.date.year:
+
+            # circumcision stops in 10-14 year olds
+            if ((self.circ_inc_rate_year_i == 1)
+                | (self.circ_inc_rate_year_i == 3)
+                | (self.circ_inc_rate_year_i == 4)) \
+               & (self.year_interv <= self.date.year):
+                uncirc_male_population = pop.data.index[(pop.data[col.SEX] == SexType.Male)
+                                                        & ~pop.data[col.CIRCUMCISED]
+                                                        & (pop.data[col.AGE] >= 15)
+                                                        & (pop.data[col.AGE] < 50)]
+            # get uncircumcised male population of specific ages
+            else:
+                uncirc_male_population = pop.data.index[(pop.data[col.SEX] == SexType.Male)
+                                                        & ~pop.data[col.CIRCUMCISED]
+                                                        & (pop.data[col.AGE] >= 10)
+                                                        & (pop.data[col.AGE] < 50)]
+
+            # continue if uncircumcised males are present this timestep
+            if len(uncirc_male_population) > 0:
+
+                # group males by age groups
+                # TODO: change age group col name to be more descriptive
+                pop.data.loc[uncirc_male_population,
+                             col.AGE_GROUP] = np.digitize(pop.data.loc[uncirc_male_population,
+                                                                       col.AGE], [10, 20, 30, 50])
+                # calculate vmmc outcomes
+                circumcision = pop.transform_group([col.AGE_GROUP], self.calc_prob_circ,
+                                                   sub_pop=uncirc_male_population)
+                pop.data.loc[uncirc_male_population, col.CIRCUMCISED] = circumcision
+                pop.data.loc[uncirc_male_population, col.VMMC] = circumcision
+                # newly circumcised males get the current date set as their circumcision date
+                new_circ_males = pop.data.index[pop.data[col.CIRCUMCISED]
+                                                & pop.data[col.CIRCUMCISION_DATE].isnull()]
+                pop.data.loc[new_circ_males, col.CIRCUMCISION_DATE] = self.date
+
+    def calc_prob_circ(self, age_group, size):
+        """
+        Calculates the circumcision probability for a given
+        age group and returns VMMC outcomes.
+        """
+        age_mod = 1
+        # age group 1 has no modifier in most cases
+        if age_group == 2:
+            age_mod = self.circ_red_20_30
+        elif age_group == 3:
+            age_mod = self.circ_red_30_50
+
+        calc_date = self.date.year
+        # cap date at 2019 for calculations
```

Comment:
Do we want 2019 and 2013 to be hardcoded?

Reply:
I don't think it matters for 2013 since that value will always stay where it is unless the way the model works is changed (e.g. if `rel_incr_circ_post_2013` is no longer relevant), but I could give both of these years variable names and put them in the yaml file for separation's sake.